ports/rp2: Per-board linker scripts to avoid producing self-corrupting builds.

Gadgetoid opened the pull request:
Add per-board linker scripts to avoid silently overflowing flash.

Excludes region allocated to user filesystem from FLASH section.

Rationale explained in https://github.com/micropython/micropython/issues/8680

TLDR: Prevents the accidental generation of MicroPython builds which self-destruct on first run by overwriting themselves with the configured LittleFS filesystem.

I'm not super happy about copying memmap_mp.ld for every board, but I don't really see a better way to do this and preprocessing is non-trivial CMake spaghetti.

Gadgetoid commented:
Note; I have created custom linker scripts for PICO and our boards which I know the flash sizes of. If this stands a chance of being merged before the next release I could try and find the time to update the other boards.

Right now this is a very us-packing-everything-and-the-kitchen-sink-into-MicroPython problem, but I think having linker scripts that correctly reflect the actual FLASH available (and not allocated to the user filesystem) makes sense generally.

dpgeorge commented:
> I'm not super happy about copying memmap_mp.ld for every board, but I don't really see a better way to do this and preprocessing is non-trivial CMake spaghetti.

It's definitely not ideal.  Would be best to try and eliminate duplication.  Linker scripts can include other linker script files, see eg `ports/stm32/boards/common_bss_heap_stack.ld` and related ld files.

You can also inject variables into linker scripts, see eg `ports/nrf/boards/memory.ld`.  That might be the best way to do it (have a variable for the flash size, which is set in the cmake file...?).

Gadgetoid commented:
Updated to a much cleaner solution.

The `pico_set_linker_script` function does not permit multiple linker scripts, so the variable approach that works elsewhere doesn't quite work as gracefully here.

In lieu of this I have detected `mpconfigboard.ld` and swapped out to it, that file is then responsible (assuming the user doesn't want to write their own complete linker script) for including `memmap_mp.ld`. The linker is given the port directory so this can be done easily with just `INCLUDE memmap_mp.ld`.

It feels a bit icky that we're defining both `MICROPY_HW_FLASH_STORAGE_BYTES` *and* `_user_fs_size`.

TODO

* [ ] Agree on naming convention for linker file, I have gone with `mpconfigboard.ld`
* [ ] Settle on a naming convention for variables describing user filesystem and flash section sizes
* [ ] Configure all other boards with their flash sizes

dpgeorge commented:
In stm32 the flash layout (what's for the bootloader, the app firmware, the filesystem etc) is done in the linker script.  And then those values are passed up to C.  I think it makes good sense to do a similar thing for the rp2 port, define everything at the linker level.

In stm32 an example is:
```
MEMORY
{
    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 1024K /* entire flash */
    FLASH_ISR (rx)  : ORIGIN = 0x08000000, LENGTH = 16K /* sector 0 */
    FLASH_FS (rx)   : ORIGIN = 0x08004000, LENGTH = 112K /* sectors 1,2,3,4 are for filesystem */
    FLASH_TEXT (rx) : ORIGIN = 0x08020000, LENGTH = 896K /* sectors 5,6,7,8,9,10,11 */
    CCMRAM (xrw)    : ORIGIN = 0x10000000, LENGTH = 64K
    RAM (xrw)       : ORIGIN = 0x20000000, LENGTH = 128K
}
```
Sometimes `FLASH_APP` is also defined for boards that use a bootloader.

I think we should use similar names for rp2, eg `FLASH` for the whole flash, `FLASH_FS` for the filesystem region, and `FLASH_APP` (or `FLASH_TEXT`) for the firmware region.

> It feels a bit icky that we're defining both `MICROPY_HW_FLASH_STORAGE_BYTES` _and_ `_user_fs_size`

I agree.  So we would remove `MICROPY_HW_FLASH_STORAGE_BYTES` and let it be set automatically by the linker/compiler.  Eg we can define this in `mpconfigboard.mk` and then use `-D...` to pass it through to C, and `--defsym=_user_fs_size=$(MICROPY_HW_FLASH_STORAGE_BYTES)` to pass it through to the linker.

Gadgetoid commented:
Your suggested approach has the consequence of making it easier to locate objects in the filesystem - ie an initialized filesystem image - at compile time. That could prove a handy jumping off point for some other ideas floating around. I like it!

Gadgetoid commented:
Okay I have replaced the ugly preprocessor hack with CMake's `configure_file` which is likely to be much, much less fragile and potentially allows for more config vars to make it through to the linker script in future without having to modify a spaghetti custom build command.

I still don't like the duplication of `PICO_FLASH_SIZE_BYTES` to `MICROPY_HW_FLASH_SIZE_BYTES` but I don't see an easy way around this. The Pico SDK toolchain preprocesses these header files and amalgamates them into one configuration header (as far as I can tell). Using the...slightly tenuous preprocessor solution I have just replaced would potentially let us `#include` the Pico SDK config but that feels backwards.

On the upside, having both these vars configured by MicroPython's board config means they're right there under your nose if you ever want to tweak the app sizes for existing boards.

dpgeorge commented:
Would it help to adjust the code in `rp2_flash.c` so that it used linker symbols to find the extents of the filesystem, rather than constants in `rp2_flash_obj`?  This is how stm32 does it: https://github.com/micropython/micropython/blob/master/ports/stm32/flashbdev.c#L39-L61

Gadgetoid commented:
The main issue with adjusting rp2_flash is the inability to use the “dynamic” linker symbols in the “bi_decl” macro. I don’t see an obvious way around this, though that may be a lack of knowledge or imagination on my part.

Gadgetoid commented:
Apart from this not being marked "Ready for review" is there anything that needs to be addressed?

Getting input is tricky here because I'm probably the only person screaming out for this :laughing:

Gadgetoid commented:
It looks like this commit https://github.com/micropython/micropython/commit/c80e7c14e6305e50e3b39f97172d4d8fe1214d3b takes another step toward relying on `memmap_mp.ld` as a source of truth for board memory layout.

As I mentioned in the comment there, this causes some difficulty with modules that - up until now - have been able to safely assume that *some* C++ heap is available for general use. This is no longer generally the case.

Of course I'm always going to err on the side of not wanting to update our few dozen modules and expect push-back, but I think th is is another potential argument for this per-board `memmap_mp.ld` support. It would make it easier for us to ship interim MicroPython builds that leave some RAM available for our modules while we figure out how to "fix" them properly.

It's going to take some amount of effort to rebase this PR, so before I go launching into it-

1. What can we do about `bi_decl` not being compatible with dynamic linker symbols? The suggestion to make the .ld file the source of truth seems reasonable, but is incompatible with Pico SDK's binary descriptor macros. Does anyone know if there's a reasonable way to work around this?  - the problem code is https://github.com/micropython/micropython/blob/668a7bd28a49980b239fd7666684885382526988/ports/rp2/rp2_flash.c#L61-L70

2. Are there any other blockers here, other than this being a pretty much only-I-seem-to-want-this change?

Gadgetoid commented:
It should be noted that I've switched away from trying to wrangle linker scripts and have instead written a pure Python tool to analyze the .uf2 output files and check the binary metadata doesn't describe a filesystem that overlaps the flash binary.

That tool lives here: https://github.com/gadgetoid/py_decl

Is invoked in our CI like so:

```
python3 py_decl/py_decl.py --to-json --verify build-${{ matrix.name }}/${{ env.RELEASE_FILE }}.uf2
```

And will produce an error like:

```
Run python3 py_decl/py_decl.py --to-json --verify build-pico/pico-6cbf0fa0654672e92c30b670069e87383523d95c-pimoroni-micropython.uf2
CRITICAL ERROR: Block device / binary overlap!
Binary ends at 0x100a0560, block device starts at 0x100a0000
```

Gadgetoid committed:
ports/rp2: Per-board linker flash config.

Add per-board flash config to avoid silently
overwriting user code on vfs init.

Make linker FLASH section reflect board size.

Add linker APP section for user code.

Add linker FILESYSTEM section for vfs.

Add per-board config for FLASH + APP sizes:

* MICROPY_HW_FLASH_SIZE_BYTES
* MICROPY_HW_APP_SIZE_BYTES

CMake "configure_file" memmap_mp.ld.in.

Signed-off-by: Phil Howard <github@gadgetoid.com>

Gadgetoid committed:
ports/rp2: Add print-memory-usage.

Add --print-memory-usage to print memory usage at the end
of a build. Shows used/remaining app storage.

Signed-off-by: Phil Howard <github@gadgetoid.com>

Gadgetoid commented:
It's still not ideal and I'm not dealing with romfs well, but I've rebased this PR so we can better see how feasible this is and what the challenges might be.

The issues with `bi_decl` persist, since they require compile time constants and the linker symbols are only known at *link* time. This is reflected in the singleton definitions of both flash and romfs, which also require compile time constants- I've... hacked... around these.

Note: should probably have a `bi_decl` for romfs where it's included?

Gadgetoid commented:
Closed in favour of https://github.com/micropython/micropython/pull/19225
